File: teams_bot.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Integration.AspNet.Core;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

public class TeamsBot : ActivityHandler
{
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly string _apiBaseUrl;

    public TeamsBot(IHttpClientFactory httpClientFactory, IConfiguration configuration)
    {
        _httpClientFactory = httpClientFactory;
        _apiBaseUrl = configuration["ApiBaseUrl"];
    }

    protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
    {
        var userInput = turnContext.Activity.Text;

        if (userInput.StartsWith("upload"))
        {
            // Handle document upload
            var fileData = await DownloadFileAsync(turnContext.Activity.Attachments[0].ContentUrl);
            var extractionResult = await ExtractTextAsync(fileData);
            await turnContext.SendActivityAsync(MessageFactory.Text("Document uploaded and processed successfully."));
            await turnContext.SendActivityAsync(MessageFactory.Text($"Extracted text: {extractionResult["chunks"][0]}..."));
        }
        else if (userInput.StartsWith("question"))
        {
            // Handle question submission
            var question = userInput.Substring("question ".Length);
            var summaries = new Dictionary<string, string>(); // Fetch summaries from your storage
            var relevantDoc = await SelectRelevantAsync(question, summaries);
            var answer = await GetAnswerAsync(question, relevantDoc["most_relevant"]);
            await turnContext.SendActivityAsync(MessageFactory.Text($"Answer: {answer}"));
        }
        else
        {
            await turnContext.SendActivityAsync(MessageFactory.Text("Please upload a document or ask a question."));
        }
    }

    private async Task<byte[]> DownloadFileAsync(string fileUrl)
    {
        var client = _httpClientFactory.CreateClient();
        return await client.GetByteArrayAsync(fileUrl);
    }

    private async Task<Dictionary<string, object>> ExtractTextAsync(byte[] fileData)
    {
        var client = _httpClientFactory.CreateClient();
        using var content = new MultipartFormDataContent();
        content.Add(new ByteArrayContent(fileData), "file", "document.pdf");
        var response = await client.PostAsync($"{_apiBaseUrl}/extract_text/", content);
        response.EnsureSuccessStatusCode();
        var responseString = await response.Content.ReadAsStringAsync();
        return JsonConvert.DeserializeObject<Dictionary<string, object>>(responseString);
    }

    private async Task<Dictionary<string, object>> SelectRelevantAsync(string question, Dictionary<string, string> summaries)
    {
        var client = _httpClientFactory.CreateClient();
        var content = new StringContent(JsonConvert.SerializeObject(new { question, summaries }), System.Text.Encoding.UTF8, "application/json");
        var response = await client.PostAsync($"{_apiBaseUrl}/select_relevant/", content);
        response.EnsureSuccessStatusCode();
        var responseString = await response.Content.ReadAsStringAsync();
        return JsonConvert.DeserializeObject<Dictionary<string, object>>(responseString);
    }

    private async Task<string> GetAnswerAsync(string question, string documentText)
    {
        var client = _httpClientFactory.CreateClient();
        var content = new StringContent(JsonConvert.SerializeObject(new { question, documentText }), System.Text.Encoding.UTF8, "application/json");
        var response = await client.PostAsync($"{_apiBaseUrl}/get_answer/", content);
        response.EnsureSuccessStatusCode();
        var responseString = await response.Content.ReadAsStringAsync();
        return JsonConvert.DeserializeObject<Dictionary<string, string>>(responseString)["answer"];
    }
}

public class AdapterWithErrorHandler : BotFrameworkHttpAdapter
{
    public AdapterWithErrorHandler(IConfiguration configuration, ILogger<BotFrameworkHttpAdapter> logger)
        : base(configuration, logger)
    {
        OnTurnError = async (turnContext, exception) =>
        {
            logger.LogError($"Exception caught : {exception}");
            await turnContext.SendActivityAsync(MessageFactory.Text("Sorry, it looks like something went wrong."));
        };
    }
}

public class Startup
{
    public IConfiguration Configuration { get; }

    public Startup(IConfiguration configuration)
    {
        Configuration = configuration;
    }

    public void ConfigureServices(IServiceCollection services)
    {
        services.AddHttpClient();
        services.AddControllers().AddNewtonsoftJson();
        services.AddSingleton<IBotFrameworkHttpAdapter, AdapterWithErrorHandler>();
        services.AddTransient<IBot, TeamsBot>();
    }

    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        if (env.IsDevelopment())
        {
            app.UseDeveloperExceptionPage();
        }

        app.UseDefaultFiles();
        app.UseStaticFiles();
        app.UseRouting();
        app.UseAuthorization();
        app.UseEndpoints(endpoints =>
        {
            endpoints.MapControllers();
        });
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        CreateHostBuilder(args).Build().Run();
    }

    public static IHostBuilder CreateHostBuilder(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureWebHostDefaults(webBuilder =>
            {
                webBuilder.UseStartup<Startup>();
            });
}

File: teams_bot.py
import os
from aiohttp import web
from botbuilder.core import (
    BotFrameworkAdapterSettings,
    TurnContext,
    BotFrameworkAdapter,
    ConversationState,
    MemoryStorage,
)
from botbuilder.schema import Activity, ActivityTypes
from services.document_service import (
    count_tokens,
    split_text,
    extract_text,
    summarize,
    process_chunks,
    select_relevant,
    get_answer,
)

# Bot configuration
APP_ID = os.getenv("MICROSOFT_APP_ID", "")
APP_PASSWORD = os.getenv("MICROSOFT_APP_PASSWORD", "")
adapter_settings = BotFrameworkAdapterSettings(APP_ID, APP_PASSWORD)
adapter = BotFrameworkAdapter(adapter_settings)

# Memory storage for conversation state
memory_storage = MemoryStorage()
conversation_state = ConversationState(memory_storage)


async def handle_message(context: TurnContext):
    if context.activity.type == ActivityTypes.message:
        user_input = context.activity.text

        if user_input.startswith("upload"):
            # Handle document upload
            file_data = await context.activity.attachments[0].download()
            extraction_result = await extract_text(file_data)
            await context.send_activity("Document uploaded and processed successfully.")
            await context.send_activity(f"Extracted text: {extraction_result['chunks'][:500]}...")

        elif user_input.startswith("question"):
            # Handle question submission
            question = user_input[len("question "):]
            summaries = {}  # Fetch summaries from your storage
            relevant_doc = await select_relevant(question, summaries)
            answer = await get_answer(question, relevant_doc["most_relevant"])
            await context.send_activity(f"Answer: {answer}")

        else:
            await context.send_activity("Please upload a document or ask a question.")

    await conversation_state.save_changes(context)


async def messages(req: web.Request) -> web.Response:
    body = await req.json()
    activity = Activity().deserialize(body)
    auth_header = req.headers.get("Authorization", "")

    response = await adapter.process_activity(activity, auth_header, handle_message)
    return web.json_response(data=response.body, status=response.status)


app = web.Application()
app.router.add_post("/api/messages", messages)

if __name__ == "__main__":
    web.run_app(app, host="localhost", port=3978)
